### PR TITLE
fix: topk fp32 chunk memory

### LINKED ISSUE #2882: [bug] `logprob_chunk_size` is broken in distillation and can cause OOM
**Describe the bug**

`logprob_chunk_size` does not bound peak GPU memory in `distributed_vocab_topk`
and `gather_logits_at_global_indices` (`nemo_rl/distributed/model_utils.py`) in distillation.
Both functions upcast the entire `[B, S, V_local]` logits tensor to fp32
**before** the per-chunk loop, so the fp32 working set scales with the full
sequence length `S` instead of `chunk_size`. The chunking therefore gives no
memory benefit, and long sequences (e.g. distillation at 32K) hit avoidable OOMs.

The offending line (hoisted above the loop) in each function:
- [`distributed_vocab_topk`](https://github.com/NVIDIA-NeMo/RL/blob/29fc948c5debe13373616b56661ab164696dec6c/nemo_rl/distributed/model_utils.py#L1452)
- [`gather_logits_at_global_indices`](https://github.com/NVIDIA-NeMo/RL/blob/29fc948c5debe13373616b56661ab164696dec6c/nemo_rl/distributed/model_utils.py#L1541)

```python
logits = vocab_parallel_logits.to(dtype=torch.float32)  # whole [B, S, V_local] in fp32, before chunking
```

**Steps/Code to reproduce bug**

```python
import torch
from nemo_rl.distributed.model_utils import gather_logits_at_global_indices

B, S, V_local, k = 1, 32768, 16032, 64
logits = torch.randn(B, S, V_local, device="cuda", dtype=torch.bfloat16)
idx = torch.randint(0, V_local, (B, S, k), device="cuda")

for chunk in (None, 512):
    torch.cuda.reset_peak_memory_stats()
    gather_logits_at_global_indices(
        logits, idx, tp_group=None,
        vocab_start_index=0, vocab_end_index=V_local, chunk_size=chunk,
    )
    print(chunk, f"{torch.cuda.max_memory_allocated() / 1e6:.0f} MB")
# Expected: chunk_size=512 memory should be lower than chunk_size=None.
# Actual (B200, bf16): None -> 3213 MB, 512 -> 3186 MB (chunk_size ignored).
```

**Expected behavior**

Peak fp32 memory should scale with `chunk_size`, not the full sequence length,
so that `logprob_chunk_size` can bound memory for long-sequence distillation.

**Additional context**

### PR DESCRIPTION
**Pull request by odedovadia**
# What does this PR do ?

**Upcast logits to fp32 one sequence-chunk at a time in `distributed_vocab_topk`
and `gather_logits_at_global_indices`, so `logprob_chunk_size` actually bounds
the fp32 working set instead of materializing the whole `[B, S, V_local]` tensor in distillation.**

# Issues

Closes #2882 

# Usage

Set `logprob_chunk_size` in the policy config for distillation:

```yaml
policy:
  logprob_chunk_size: 512   # bounds fp32 peak to [B, 512, V_local]
  defer_fp32_logits: true
```

# Memory analysis

Peak CUDA memory of the pre-fix vs post-fix at the same `chunk_size=512`, as measured on one B200 with bf16 logits, `B=1`, `V_local=128256` (llama3 vocab, TP=1), `k=64`. 

Pre-fix is very large and grows with seq_len, while post-fix is bounded by the chunk size. 

| seq_len | Pre-fix memory peak | Post-fix memory peak | Reduction |
|--------:|----------:|------------------:|----------:|
|    8192 |  4055 MiB |           508 MiB |    87.4%  |
|   16384 |  8109 MiB |           514 MiB |    93.6%  |
|   32768 | 16219 MiB |           526 MiB |    96.7%  |

`gather_logits_at_global_indices` behaves identically. At `chunk_size=None` both implementations match exactly as expected.

# Before your PR is "Ready for review"

**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

* Added equivalence tests in `tests/unit/distributed/test_model_utils.py`
  (`test_distributed_vocab_topk_ops`).
* Outputs are identical before/after the change - only peak memory differs.
* This aligns the two ops with the existing `ChunkedDistributedLogprob` (GRPO  logprob path), which already upcasts to fp32 inside the chunk loop.

### CODE REVIEW

**Comment by RayenTian** [comment]
/ok to test [87a5c9d](https://github.com/NVIDIA-NeMo/RL/pull/2883/commits/87a5c9d437269d62fef291094c354023500e8e92)

**Comment by RayenTian** [comment]
Thanks for the clean repro and fix, @odedovadia! 
The change is exactly right.  LGTM pending CI. Would you please also take a look @yuki-97.

**Comment by RayenTian** [comment]
@odedovadia Would you please solve the lint check failed? You can refer to this [link](https://github.com/NVIDIA-NeMo/RL/pull/2210#issuecomment-4240855901) for how to solve it.

**Comment by odedovadia** [comment]
> @odedovadia Would you please solve the lint check failed? You can refer to this [link](https://github.com/NVIDIA-NeMo/RL/pull/2210#issuecomment-4240855901) for how to solve it.

Fixed :)

**Comment by RayenTian** [comment]
/ok to test [31b82d3](https://github.com/NVIDIA-NeMo/RL/pull/2883/commits/31b82d3d08dc3c1b3446c211df4fc0cc2ceb8527)

**Review by yuki-97 — APPROVED** [review]
@odedovadia  LGTM, thanks for the fix!
